File: LightWave2D/pml.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from pydantic.dataclasses import dataclass
from TypedUnit import AnyUnit, ureg

from LightWave2D.grid import Grid
from LightWave2D.utils import config_dict


@dataclass(config=config_dict, kw_only=True)
class PML:
    """
    Represents a Perfectly Matched Layer (PML) for absorbing boundary conditions in FDTD simulations.

    Parameters
    ----------
    grid : Grid
        The simulation grid.
    width : str
        Width of the PML region in grid cells, expressed as a percentage (e.g., "10%").
    sigma_max : AnyUnit
        Maximum value of the conductivity profile.
    order : int
        Polynomial order of the conductivity profile.
    """
    grid: Grid
    width: str = "10"
    sigma_max: AnyUnit = 0.045 * ureg.siemens / ureg.meter
    order: int = 3

    def __post_init__(self):
        """
        Initialize the PML regions with appropriate conductivity profiles.
        """
        self.sigma_x = np.zeros((self.grid.n_x, self.grid.n_y))
        self.sigma_y = np.zeros((self.grid.n_x, self.grid.n_y))

        y_mesh, x_mesh = np.meshgrid(self.grid.y_stamp, self.grid.x_stamp)

        opposite_width = float(self.width.strip("%"))
        opposite_width = f"{100 - opposite_width}%"
        self.width_start = self.grid.get_coordinate(x=self.width, y=self.width)
        self.width_stop = self.grid.get_coordinate(x=opposite_width, y=opposite_width)

        bottom_boundary = y_mesh < self.width_start.y
        left_boundary = x_mesh < self.width_start.x

        top_boundary = y_mesh > self.width_stop.y
        right_boundary = x_mesh > self.width_stop.x

        _sigma_max = self.sigma_max.to('siemens/meter').magnitude

        self.sigma_y[bottom_boundary] = _sigma_max * ((self.width_start.y.to('meter').magnitude - y_mesh[bottom_boundary].to('meter').magnitude) / self.width_start.y.to('meter').magnitude) ** self.order
        self.sigma_y[top_boundary] = _sigma_max * ((y_mesh[top_boundary].to('meter').magnitude - self.width_stop.y.to('meter').magnitude) / self.width_start.y.to('meter').magnitude) ** self.order
        self.sigma_x[left_boundary] = _sigma_max * ((self.width_start.x.to('meter').magnitude - x_mesh[left_boundary].to('meter').magnitude) / self.width_start.x.to('meter').magnitude) ** self.order
        self.sigma_x[right_boundary] = _sigma_max * ((x_mesh[right_boundary].to('meter').magnitude - self.width_stop.x.to('meter').magnitude) / self.width_start.x.to('meter').magnitude) ** self.order

        self.sigma_x *= ureg.siemens / ureg.meter
        self.sigma_y *= ureg.siemens / ureg.meter

    def add_to_ax(self, ax: plt.Axes, distance_units: ureg.Quantity = ureg.meter) -> None:
        """
        Add the PML regions to a matplotlib axis.

        Parameters
        ----------
        ax : plt.Axes
            The axis to which the PML regions will be added.
        """
        cmap = np.zeros([256, 4])
        cmap[:, 3] = np.linspace(0, 1, 256)
        cmap = ListedColormap(cmap)

        ax.pcolormesh(
            self.grid.x_stamp.to('meter').magnitude,
            self.grid.y_stamp.to('meter').magnitude,
            self.sigma_y.T.to('siemens/meter').magnitude + self.sigma_x.T.to('siemens/meter').magnitude,
            cmap=cmap,
        )

    def plot(self, unit_size: int = 6) -> None:
        """
        Plot the PML regions.

        Parameters
        ----------
        unit_size : int, optional
            Size of each unit in the plot (default is 6).
        """
        figsize = int(unit_size), int(unit_size * self.grid.size_y / self.grid.size_x)
        figure, ax = plt.subplots(1, 1, figsize=figsize)

        ax.set_title('FDTD Simulation PML Regions')
        ax.set_ylabel(r'y position [$\mu$m]')
        ax.set_xlabel(r'x position [$\mu$m]')
        ax.set_aspect('equal')

        self.add_to_ax(ax)
        plt.show()

File: LightWave2D/cpp/pml/pml.cpp
#include "pml.h"

#include <cmath>
#include <string>

#include <pybind11/numpy.h>

namespace py = pybind11;

namespace {
double meter(const py::object& value) { return value.attr("to")("meter").attr("magnitude").cast<double>(); }
double conductivity(const py::object& value) { return value.attr("to")("siemens/meter").attr("magnitude").cast<double>(); }
std::string percentage(py::object value) { std::string text = py::str(value).cast<std::string>(); return text.find('%') == std::string::npos ? text + "%" : text; }
}

PML::PML(py::object grid_, py::object width_, py::object sigma_max_, int order_)
    : grid(std::move(grid_)), width(std::move(width_)), sigma_max(std::move(sigma_max_)), order(order_) {
    if (order < 1) throw py::value_error("order must be at least 1.");
    const double width_percent = std::stod(percentage(width).substr(0, percentage(width).size() - 1));
    if (width_percent <= 0.0 || width_percent >= 50.0) throw py::value_error("width must be greater than 0% and less than 50%.");
    const std::string start = percentage(width);
    const std::string stop = std::to_string(100.0 - width_percent) + "%";
    width_start = grid.attr("get_coordinate")(py::arg("x") = start, py::arg("y") = start);
    width_stop = grid.attr("get_coordinate")(py::arg("x") = stop, py::arg("y") = stop);
    const int64_t nx = grid.attr("n_x").cast<int64_t>(), ny = grid.attr("n_y").cast<int64_t>();
    py::array_t<double> x({nx, ny}), y({nx, ny}); auto xv = x.mutable_unchecked<2>(); auto yv = y.mutable_unchecked<2>();
    const py::array_t<double> x_stamp = grid.attr("x_stamp").attr("to")("meter").attr("magnitude").cast<py::array_t<double>>();
    const py::array_t<double> y_stamp = grid.attr("y_stamp").attr("to")("meter").attr("magnitude").cast<py::array_t<double>>();
    const auto xs = x_stamp.unchecked<1>();
    const auto ys = y_stamp.unchecked<1>();
    const double start_x = meter(width_start.attr("x")), start_y = meter(width_start.attr("y"));
    const double stop_x = meter(width_stop.attr("x")), stop_y = meter(width_stop.attr("y"));
    const double maximum = conductivity(sigma_max);
    for (int64_t i = 0; i < nx; ++i) for (int64_t j = 0; j < ny; ++j) {
        const double left = xs(i) < start_x ? std::pow((start_x - xs(i)) / start_x, order) : 0.0;
        const double right = xs(i) > stop_x ? std::pow((xs(i) - stop_x) / start_x, order) : 0.0;
        const double bottom = ys(j) < start_y ? std::pow((start_y - ys(j)) / start_y, order) : 0.0;
        const double top = ys(j) > stop_y ? std::pow((ys(j) - stop_y) / start_y, order) : 0.0;
        xv(i, j) = maximum * (left + right); yv(i, j) = maximum * (bottom + top);
    }
    py::object unit = py::module_::import("TypedUnit").attr("ureg").attr("siemens") / py::module_::import("TypedUnit").attr("ureg").attr("meter");
    sigma_x = x * unit; sigma_y = y * unit;
}

void PML::add_to_ax(const py::object& axis) const {
    py::object colors = py::module_::import("matplotlib.colors"); py::object numpy = py::module_::import("numpy");
    py::object rgba = numpy.attr("zeros")(py::make_tuple(256, 4)); rgba.attr("__setitem__")(py::make_tuple(py::ellipsis(), 3), numpy.attr("linspace")(0, 1, 256));
    py::object cmap = colors.attr("ListedColormap")(rgba);
    py::object x = grid.attr("x_stamp").attr("to")("meter").attr("magnitude");
    py::object y = grid.attr("y_stamp").attr("to")("meter").attr("magnitude");
    py::object conductivity = sigma_y.attr("to")("siemens/meter").attr("magnitude") + sigma_x.attr("to")("siemens/meter").attr("magnitude");
    py::object transposed_conductivity = conductivity.attr("T");
    axis.attr("pcolormesh")(x, y, transposed_conductivity, py::arg("cmap") = cmap);
}

void PML::plot(int unit_size) const { py::object pyplot=py::module_::import("matplotlib.pyplot");py::tuple items=pyplot.attr("subplots")(1,1,py::arg("figsize")=py::make_tuple(unit_size,unit_size));py::object axis=items[1];axis.attr("set_aspect")("equal");add_to_ax(axis);pyplot.attr("show")(); }
